src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -60,29 +62,164 @@ protected override void Dispose(bool finalize)
+        ///         it let that request to update primary solution cache to latest.
+        /// 
+        /// these 2 are complimentary to each other. #1 makes OOP's primary solution to be ready for next call (push), #2 makes OOP's primary
+        /// solution be not stale as much as possible. (pull)
         /// </summary>
-        public bool TryAddSolutionIfPossible(SolutionInfo solutionInfo, int workspaceVersion, SerializableOptionSet options, [NotNullWhen(true)] out Solution? solution)
+        private async Task<Solution> CreateSolution_NoLockAsync(
+            AssetProvider assetProvider,
+            Checksum solutionChecksum,
+            bool fromPrimaryBranch,
+            int workspaceVersion,
+            Solution baseSolution,
+            CancellationToken cancellationToken)
         {
-            if (solutionInfo == null)
+            try
             {
-                throw new ArgumentNullException(nameof(solutionInfo));
+                var updater = new SolutionCreator(Services.HostServices, assetProvider, baseSolution, cancellationToken);
+
+                // check whether solution is update to the given base solution
+                if (await updater.IsIncrementalUpdateAsync(solutionChecksum).ConfigureAwait(false))
+                {
+                    // create updated solution off the baseSolution
+                    var solution = await updater.CreateSolutionAsync(solutionChecksum).ConfigureAwait(false);
+
+                    if (fromPrimaryBranch)
+                    {
+                        // if the solutionChecksum is for primary branch, update primary workspace cache with the solution
+                        return UpdateSolutionIfPossible(solution, workspaceVersion);
+                    }
+
+                    // otherwise, just return the solution
+                    return solution;
+                }
+
+                // we need new solution. bulk sync all asset for the solution first.
+                await assetProvider.SynchronizeSolutionAssetsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                // get new solution info and options
+                var (solutionInfo, options) = await assetProvider.CreateSolutionInfoAndOptionsAsync(solutionChecksum, cancellationToken).ConfigureAwait(false);
+
+                if (fromPrimaryBranch)
+                {
+                    // if the solutionChecksum is for primary branch, update primary workspace cache with new solution
+                    if (TrySetCurrentSolution(solutionInfo, workspaceVersion, options, out var solution))
+                    {
+                        return solution;
+                    }
+                }
+
+                // otherwise, just return new solution
+                var workspace = new TemporaryWorkspace(Services.HostServices, WorkspaceKind.RemoteTemporaryWorkspace, solutionInfo, options);
+                return workspace.CurrentSolution;
             }
```

Comment:
i find it interesting that this is how we create the solution, as opposed to potentially just forking hte actual primary solution to this.  but i suppose that's how we've always done this, so it might be risky to change.

Reply:
Yeah, I did not want to change the actual logic at this point.